### ISSUE TITLE
[Bug] Potential bug in `FixedNoiseMultiTaskGP`

### ISSUE THREAD

**Issue by yunboouyang** [pytorch/botorch#528]
# Potential bug in `FixedNoiseMultiTaskGP`

Hello botorch developers! I have tried a toy example for `FixedNoiseMultiTaskGP`. However I don't think results make sense to me. The setup is similar to the example in the source code comments.

We have 2 tasks. Task 1 has 1000 1-dimensional random input values while task 2 has 2000 1-dimensional random input values. The underlying function for both tasks is fixed as `sin` function in the interval `[0, 1]`. A small scale random variance for outputs is added.  Posterior means for both tasks on the test points `[0.01, 0.02, ..., 1.0]` are obtained. 

However I got the following results which indicates the posterior mean functions are decreasing (the `sin` function in the interval `[0, 1]` is increasing) and function fitting does a poor job (e.g., `sin(0.01) \approx 0.01` but not close to 0.4592). Since this example is simple enough and there should be no need to do input normalization or output standardization and sample size is large enough, I'm wondering if this indicates implementation has some bug. 

## To reproduce
** Code snippet to reproduce **
```python
torch.manual_seed(0)
X1, X2 = torch.rand(1000, 1), torch.rand(2000, 1)
i1, i2 = torch.zeros(1000, 1), torch.ones(2000, 1)
train_X = torch.cat([torch.cat([X1, i1], -1), torch.cat([X2, i2], -1)], dim=0)
train_Y = torch.cat([torch.sin(X1), torch.sin(X2)])
train_Yvar = 0.01 + 0.01 * torch.rand_like(train_Y)
model = FixedNoiseMultiTaskGP(train_X, train_Y, train_Yvar, -1)
test_parameter_value_tensor = torch.from_numpy(np.transpose(np.array([[x /100 for x in range(100)]])))
model.posterior(test_parameter_value_tensor.float()).mean
```

** Stack trace/error message **
```
posterior mean results
tensor([[0.4592, 0.4592],
        [0.4590, 0.4590],
        [0.4583, 0.4583],
        [0.4582, 0.4582],
        [0.4576, 0.4576],
        [0.4569, 0.4569],
        [0.4561, 0.4561],
        [0.4548, 0.4548],
        [0.4536, 0.4536],
        [0.4523, 0.4523],
        [0.4506, 0.4506],
        [0.4489, 0.4489],
        [0.4470, 0.4470],
        [0.4458, 0.4458],
        [0.4435, 0.4435],
        [0.4413, 0.4413],
        [0.4391, 0.4391],
        [0.4370, 0.4370],
        [0.4343, 0.4343],
        [0.4317, 0.4317],
        [0.4287, 0.4287],
        [0.4259, 0.4259],
        [0.4233, 0.4233],
        [0.4197, 0.4197],
        [0.4167, 0.4167],
        [0.4135, 0.4135],
        [0.4107, 0.4107],
        [0.4072, 0.4072],
        [0.4037, 0.4037],
        [0.4000, 0.4000],
        [0.3963, 0.3963],
        [0.3929, 0.3929],
        [0.3884, 0.3884],
        [0.3855, 0.3855],
        [0.3812, 0.3812],
        [0.3779, 0.3779],
        [0.3735, 0.3735],
        [0.3700, 0.3700],
        [0.3658, 0.3658],
        [0.3617, 0.3617],
        [0.3577, 0.3577],
        [0.3533, 0.3533],
        [0.3498, 0.3498],
        [0.3452, 0.3452],
        [0.3411, 0.3411],
        [0.3368, 0.3368],
        [0.3328, 0.3328],
        [0.3287, 0.3287],
        [0.3245, 0.3245],
        [0.3203, 0.3203],
        [0.3162, 0.3162],
        [0.3118, 0.3118],
        [0.3076, 0.3076],
        [0.3035, 0.3035],
        [0.2993, 0.2993],
        [0.2952, 0.2952],
        [0.2911, 0.2911],
        [0.2871, 0.2871],
        [0.2824, 0.2824],
        [0.2786, 0.2786],
        [0.2746, 0.2746],
        [0.2704, 0.2704],
        [0.2663, 0.2663],
        [0.2626, 0.2626],
        [0.2583, 0.2583],
        [0.2541, 0.2541],
        [0.2503, 0.2503],
        [0.2464, 0.2464],
        [0.2426, 0.2426],
        [0.2386, 0.2386],
        [0.2345, 0.2345],
        [0.2307, 0.2307],
        [0.2270, 0.2270],
        [0.2233, 0.2233],
        [0.2195, 0.2195],
        [0.2158, 0.2158],
        [0.2121, 0.2121],
        [0.2086, 0.2086],
        [0.2049, 0.2049],
        [0.2013, 0.2013],
        [0.1980, 0.1980],
        [0.1944, 0.1944],
        [0.1910, 0.1910],
        [0.1877, 0.1877],
        [0.1844, 0.1844],
        [0.1808, 0.1808],
        [0.1778, 0.1778],
        [0.1743, 0.1743],
        [0.1712, 0.1712],
        [0.1682, 0.1682],
        [0.1650, 0.1650],
        [0.1618, 0.1618],
        [0.1588, 0.1588],
        [0.1559, 0.1559],
        [0.1529, 0.1529],
        [0.1501, 0.1501],
        [0.1473, 0.1473],
        [0.1443, 0.1443],
        [0.1415, 0.1415],
        [0.1390, 0.1390]], grad_fn=<CopyBackwards>)
```

## Expected Posterior Mean Value
```
torch.sin(test_parameter_value_tensor)
tensor([[0.0000],
        [0.0100],
        [0.0200],
        [0.0300],
        [0.0400],
        [0.0500],
        [0.0600],
        [0.0699],
        [0.0799],
        [0.0899],
        [0.0998],
        [0.1098],
        [0.1197],
        [0.1296],
        [0.1395],
        [0.1494],
        [0.1593],
        [0.1692],
        [0.1790],
        [0.1889],
        [0.1987],
        [0.2085],
        [0.2182],
        [0.2280],
        [0.2377],
        [0.2474],
        [0.2571],
        [0.2667],
        [0.2764],
        [0.2860],
        [0.2955],
        [0.3051],
        [0.3146],
        [0.3240],
        [0.3335],
        [0.3429],
        [0.3523],
        [0.3616],
        [0.3709],
        [0.3802],
        [0.3894],
        [0.3986],
        [0.4078],
        [0.4169],
        [0.4259],
        [0.4350],
        [0.4439],
        [0.4529],
        [0.4618],
        [0.4706],
        [0.4794],
        [0.4882],
        [0.4969],
        [0.5055],
        [0.5141],
        [0.5227],
        [0.5312],
        [0.5396],
        [0.5480],
        [0.5564],
        [0.5646],
        [0.5729],
        [0.5810],
        [0.5891],
        [0.5972],
        [0.6052],
        [0.6131],
        [0.6210],
        [0.6288],
        [0.6365],
        [0.6442],
        [0.6518],
        [0.6594],
        [0.6669],
        [0.6743],
        [0.6816],
        [0.6889],
        [0.6961],
        [0.7033],
        [0.7104],
        [0.7174],
        [0.7243],
        [0.7311],
        [0.7379],
        [0.7446],
        [0.7513],
        [0.7578],
        [0.7643],
        [0.7707],
        [0.7771],
        [0.7833],
        [0.7895],
        [0.7956],
        [0.8016],
        [0.8076],
        [0.8134],
        [0.8192],
        [0.8249],
        [0.8305],
        [0.8360]], dtype=torch.float64)
```

## System information

**Please complete the following information:**
-  botorch version 0.3.0
-  gpytorch version 1.1.1
-  PyTorch version 1.6.0
-  Mac OS

**Comment by yunboouyang** [comment]
I switched to `MultiTaskGP`. The performance is still bad. I shared the code as follows:
```
torch.manual_seed(0)
X1, X2 = torch.rand(1000, 1), torch.rand(2000, 1)
i1, i2 = torch.zeros(1000, 1), torch.ones(2000, 1)
train_X = torch.cat([torch.cat([X1, i1], -1), torch.cat([X2, i2], -1)], dim=0)
train_Y = torch.cat([torch.sin(X1), torch.sin(X2)])
model = MultiTaskGP(train_X, train_Y,  -1)
test_parameter_value_tensor = torch.from_numpy(np.transpose(np.array([[x /100 for x in range(100)]])))
model.posterior(test_parameter_value_tensor.float()).mean
```

**Comment by Balandat** [comment]
@yunboouyang Are you actually fitting the model? From your code it looks like you are not. 

Try 
```
from gpytorch.mlls import ExactMarginalLogLikelihood
from botorch.fit import fit_gpytorch_model
mll = ExactMarginalLogLikelihood(model.likelihood, model)
```

and then compute predictions. 

*Note*: 2K + 1K data points is quite a lot for an exact multi-task GP, this will take a little while to fit. 

Also note that we recently fixed a bug regarding the posterior variances of MultiTaskGP that hasn't made it into a release yet, you may want to work off master to get that fix.

**Comment by yunboouyang** [comment]
Thank you! I added the fitting step and reduced the sample size, but I got constant posterior mean values:
```
from botorch.models import MultiTaskGP
from gpytorch.mlls import ExactMarginalLogLikelihood
from botorch.fit import fit_gpytorch_model
import numpy as np
import torch


torch.manual_seed(0)
X1, X2 = torch.rand(50, 1), torch.rand(50, 1)
i1, i2 = torch.zeros(50, 1), torch.ones(50, 1)
train_X = torch.cat([torch.cat([X1, i1], -1), torch.cat([X2, i2], -1)], dim=0)
train_Y = torch.cat([torch.sin(X1), torch.sin(X2)])
model = MultiTaskGP(train_X, train_Y,  -1)
mll = ExactMarginalLogLikelihood(model.likelihood, model)
mll = fit_gpytorch_model(mll)
test_parameter_value_tensor = torch.from_numpy(np.transpose(np.array([[x /10 for x in range(10)]])))
model.posterior(test_parameter_value_tensor.float()).mean
```

Here I got constant posterior means
```
tensor([[0.4343, 0.4343],
        [0.4343, 0.4343],
        [0.4343, 0.4343],
        [0.4343, 0.4343],
        [0.4343, 0.4343],
        [0.4343, 0.4343],
        [0.4343, 0.4343],
        [0.4343, 0.4343],
        [0.4343, 0.4343],
        [0.4343, 0.4343]], grad_fn=<SinBackward>)
```

I followed the similar setup as unit tests in the code base https://github.com/pytorch/botorch/blob/3849c36284dc5e1559d6bc9845bdfa069de842a8/test/models/test_multitask.py#L93

**Comment by Balandat** [comment]
Hmm I I run your code verbatim I get the following (which looks very reasonable):

```
tensor([[0.0016, 0.0015],
        [0.0993, 0.0992],
        [0.1986, 0.1985],
        [0.2957, 0.2957],
        [0.3895, 0.3895],
        [0.4794, 0.4794],
        [0.5645, 0.5646],
        [0.6441, 0.6442],
        [0.7174, 0.7175],
        [0.7834, 0.7835]], grad_fn=<CopyBackwards>)
```

You are on version 0.3.0? Can you try running on master to see if that MTGP fix is the missing ingredient? The fact that our `grad_fn` is different for the output tensor is a little concerning...

**Comment by yunboouyang** [comment]
I can confirm the botorch version is 0.3.0.
```
>>> print(botorch.__version__)
0.3.0
```

**Comment by yunboouyang** [comment]
Do you mean I should work off the master? Which branch should I work with?

**Comment by Balandat** [comment]
Yes, you can try installing from the master branch using the instructions here: https://github.com/pytorch/botorch#installing-from-latest-master

**Comment by yunboouyang** [comment]
Thank you, Max! I confirm I can get reasonable results from the latest master
```
>>> model.posterior(test_parameter_value_tensor.float()).mean
tensor([[0.0016, 0.0014],
        [0.0993, 0.0993],
        [0.1986, 0.1985],
        [0.2957, 0.2957],
        [0.3895, 0.3895],
        [0.4794, 0.4794],
        [0.5645, 0.5646],
        [0.6441, 0.6442],
        [0.7174, 0.7175],
        [0.7834, 0.7835]], grad_fn=<CopyBackwards>)
```
This is really helpful! Could I ask when you plan to release the newest code change as a new version?

**Comment by Balandat** [comment]
We don't have any major changes to merge in in the short term that would warrant a feature release, but we can probably put out a 0.3.1 bugfix release soon.

**Comment by yunboouyang** [comment]
Thank you Max! A bug fix release will also be helpful.

**Comment by Balandat** [comment]
0.3.1 that includes the fix has just been released
